Example/Tests/Controllers/MHVBrowserAuthBrokerTests.m

```diff
@@ -0,0 +1,348 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+#import <XCTest/XCTest.h>
+#import "MHVBrowserAuthBroker.h"
+#import "MHVErrorConstants.h"
+#import <WebKit/WebKit.h>
+#import "Kiwi.h"
+
+#pragma mark - Helper classes
+
+// Properties are readonly in WKNavigationAction, override request getter to return test value
+@interface MHVNavigationAction : WKNavigationAction
+
+@property (nonatomic, copy) NSURLRequest *testRequest;
+
+@end
+
+@implementation MHVNavigationAction
+
+- (NSURLRequest *)request
+{
+    return self.testRequest;
+}
+
+@end
+
+// Test browser controller that calls completion for dismiss
+@interface MHVTestBrowserController : MHVBrowserController
+
+@end
+
+@implementation MHVTestBrowserController
+
+- (void)dismissViewControllerAnimated:(BOOL)flag completion:(void (^)(void))completion
+{
+    if (completion)
+    {
+        completion();
+    }
+}
+
+@end
+
+// Access private property to set MHVTestBrowserController
+@interface MHVBrowserAuthBroker (PrivateProperties)
+
+@property (nonatomic, strong) MHVBrowserController *browserController;
+
+@end
+
+
+#pragma mark - Tests
+
+SPEC_BEGIN(MHVBrowserAuthBrokerTests)
+
+describe(@"MHVBrowserAuthBrokerTests", ^
+{
+    __block MHVBrowserAuthBroker *authBroker;
+    __block UIViewController *presentedViewController;
+    __block void (^presentedViewControllerAction)(void);
+    __block NSURL *returnedSuccessUrl;
+    __block NSError *returnedError;
+    __block WKNavigationActionPolicy returnedNavigationPolicy;
+
+    WKWebView *nilWebView = nil;
+    
+    // Mocks for view controller
+    KWMock *mockVC = [UIViewController mock];
+    
+    [mockVC stub:@selector(presentViewController:animated:completion:) withBlock:^id (NSArray *params)
+     {
+         presentedViewController = params[0];
+         
+         if (presentedViewControllerAction)
+         {
+             presentedViewControllerAction();
+         }
+         
+         // Test framework use NSNull if a parameter is nil & completion is nullable
+         if (![params[2] isKindOfClass:[NSNull class]])
+         {
+             void (^completion)(void) = params[2];
+             completion();
+         }
+         
+         return nil;
+     }];
+    
+    // Clear and create new objects before each test
+    beforeEach(^{
+        authBroker = [[MHVBrowserAuthBroker alloc] init];
+        authBroker.browserController = [[MHVTestBrowserController alloc] initWithAuthBroker:authBroker];
+        
+        presentedViewControllerAction = nil;
+        presentedViewController = nil;
+        returnedSuccessUrl = nil;
+        returnedError = nil;
+        returnedNavigationPolicy = -1;
+    });
+    
+    context(@"authenticateWithViewController", ^{
```

Comment:
When doing BDD the describe, context, and it statements should form a complete sentence e.g "when authenticateWithViewController is called". This is important for making failures clear to someone who contributes and has no knowledge of our test system and limited knowledge of our codebase. If a test fails, the contributor will see "MHVBrowserAuthBroker_whenAuthenticateWithViewControllerIsCalled_shouldPresentTheProvidedViewController." The usage of English-like sentences is a core principle of BDD.